src/systems/systems/src/hydraulic/mod.rs

```diff
@@ -1 +1,1986 @@
+                        //.linestyle("--")
+                        .linewidth(1.0),
+                )
+                .xlabel("Time [sec]")
+                .ylabel(h.name_vector[idx].as_str())
+                .legend("best")
+                .xlim(0.0, *h.time_vector.last().unwrap());
+            //.ylim(-2.0, 2.0);
+
+            curr_axis = curr_axis.grid(true);
+            all_axis.push(Some(curr_axis));
+        }
+
+        Figure::new().subplots(all_axis.len() as u32, 1, all_axis)
+    }
+
+    //History class to record a simulation
+    pub struct History {
+        time_vector: Vec<f64>,      //Simulation time starting from 0
+        name_vector: Vec<String>,   //Name of each var saved
+        data_vector: Vec<Vec<f64>>, //Vector data for each var saved
+        _data_size: usize,
+    }
+
+    impl History {
+        pub fn new(names: Vec<String>) -> History {
+            History {
+                time_vector: Vec::new(),
+                name_vector: names.clone(),
+                data_vector: Vec::new(),
+                _data_size: names.len(),
+            }
+        }
+
+        //Sets initialisation values of each data before first step
+        pub fn init(&mut self, start_time: f64, values: Vec<f64>) {
+            self.time_vector.push(start_time);
+            for v in values {
+                self.data_vector.push(vec![v]);
+            }
+        }
+
+        //Updates all values and time vector
+        pub fn update(&mut self, delta_time: f64, values: Vec<f64>) {
+            self.time_vector
+                .push(self.time_vector.last().unwrap() + delta_time);
+            self.push_data(values);
+        }
+
+        pub fn push_data(&mut self, values: Vec<f64>) {
+            for (idx, v) in values.iter().enumerate() {
+                self.data_vector[idx].push(*v);
+            }
+        }
+
+        //Builds a graph using rust crate plotlib
+        pub fn _show(self) {
+            let mut v = ContinuousView::new()
+                .x_range(0.0, *self.time_vector.last().unwrap())
+                .y_range(0.0, 3500.0)
+                .x_label("Time (s)")
+                .y_label("Value");
+
+            for cur_data in self.data_vector {
+                //Here build the 2 by Xsamples vector
+                let mut new_vector: Vec<(f64, f64)> = Vec::new();
+                for (idx, sample) in self.time_vector.iter().enumerate() {
+                    new_vector.push((*sample, cur_data[idx]));
+                }
+
+                // We create our scatter plot from the data
+                let s1: Plot = Plot::new(new_vector).line_style(LineStyle::new().colour("#DD3355"));
+
+                v = v.add(s1);
+            }
+
+            // A page with a single view is then saved to an SVG file
+            Page::single(&v).save("scatter.svg").unwrap();
+        }
+
+        //builds a graph using matplotlib python backend. PYTHON REQUIRED AS WELL AS MATPLOTLIB PACKAGE
```

Comment:
**issue**: It seems this code runs for unit tests, while creating graphs is not a unit testing concern. As I don't have some of the required dependencies installed I have output such as:

> Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'matplotlib'

**suggestion**: Please ensure all this plotting stuff doesn't execute when running `cargo test`. Consider creating a separate console application for generating plots. I'm assuming `green_loop_edp_simulation` and other functions are there to create the plots and not for actual unit testing purposes as they don't seem to have any assertion?

Reply:
Well these have assert in fact, and while plots are indeed not mandatory they provide a good way to see why a test actually failed.  Can do a separate app unless there is a way to draw graphs depending on available package?  

About the large portion not adressed... That doesn't help me much I'm afraid :/

Reply:
I understand that the graphs may be valuable for understanding a problem. However, as you're now using tests to emit them you create a situation in which you can't write good tests, as creating the graph takes far more code.

For example, this could be a test in itself, and the same goes for many other points in time like it:
> //30s, ptu should be supplying green loop

As for this:
> About the large portion not adressed... That doesn't help me much I'm afraid :/

There are a lot of unresolved items from my previous review. It's all here on this page 😉.

Reply:
> _Logic is encapsulated in types that make sense for that logic._
Not sure how to understand this?

> _Logic is further split into functions that tell the "story". Going from high-level abstraction function names all the way to lower level function names as you get deeper in the code._

A320 Logic was split in sub logic functions
Hydraulics core was also split into sub functions

> _The Law of Demeter isn't broken._

Well... That is not really specific nor helping me. 

> Avoid writing comments that explain what the code does when improving the code can tell the story without needing the comment.

Isn't always better to explain anyway?  Can we split issues from personal preferences? Not everyone that will see or modify this code is a software engineer....

> _Add an extensive collection of unit tests so changes can be made without worrying if you break something, and to make refactoring easier._

Tests have been added using your latest PR testbed improvements. I can still add tests but those already there gives a good amount of confidence both in hydraulics behaviour, and simvar intercation

Reply:
Those are the general remarks. I've sent you a message on Discord indicating what items I'm talking about. Hope that helps. Those give you examples for the questions you're asking now.

> Isn't always better to explain anyway? Can we split issues from personal preferences? Not everyone that will see or modify this code is a software engineer....

No, comments should be used to share information that cannot be clearly communicated through code. It's not a matter of personal preference but of best practice: as software developers we often see that comments go stale or become obsolete, which then makes it harder to understand the code (which one is right, the code or the comment?)

Note that not all comments are bad. That's definitely not what I'm saying. For example, the following comments are valuable:
```rust
self.left_brake_yellow_command = self.left_brake_yellow_command.min(0.5); //0.5 is temporary implementation of pressure limitation around 1000psi
self.right_brake_yellow_command = self.right_brake_yellow_command.min(0.5);
//0.5 is temporary implementation of pressure limitation around 1000psi
```
They can be further improved in terms of formatting, and by extracting a constant for the `0.5` we might even be able to remove it:
```rust
const PRESSURE_LIMITATION_AROUND_1000_PSI: f64 = 0.5;
self.left_brake_yellow_command = self.left_brake_yellow_command.min(PRESSURE_LIMITATION_AROUND_1000_PSI);
self.right_brake_yellow_command = self.right_brake_yellow_command.min(PRESSURE_LIMITATION_AROUND_1000_PSI);
```

Reply:
Still wondering why the hell I never saw those items when I first parsed your first review :(

Reply:
Test are updated with no graphics.
Still strugling to make that separate app work but that's another issue....